### ISSUE TITLE
JIT doesn't eliminate redundant branch in guarded span.Slice

### ISSUE THREAD

**Issue by stephentoub** [dotnet/runtime#65327]
[SharpLab](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIGYACMhgYRoG8aHunGAlGNgAmAeQB2AGwCeAZQAO2MQB4wAC2xQAfAwCy5ABQDh46fMUr1WhrgVi0DAJZiMjgJRcenajx+OAZgz6+gCuThiuDgxKgaHOrjaKAHQAMjBiAOYYqu7evtxeeXnEAOzWtokyEg5gMPoOrgDcHnkAvs2+JQxCMH7YwRIYTbncbcMM7fQMRqKSsrYWGto6pIaCM6bzaotlivZhbu0FhQ4BQbHhkUoAvDFh8eWpGVk5hQxHr9ydCWIVVTV1jXaPlGr063V6/UG7VGLSAA)
```C#
using System;
public class C
{
    public ReadOnlySpan<char> M1(ReadOnlySpan<char> span, int i)
    {
        if ((uint)i < (uint)span.Length)
        {
            return span.Slice(i);
        }
        return default;
    }
    
    public ReadOnlySpan<char> M2(ReadOnlySpan<char> span, int i)
    {
        if ((uint)i <= (uint)span.Length)
        {
            return span.Slice(i);
        }
        return default;
    }
}
```

```
C.M1(System.ReadOnlySpan`1<Char>, Int32)
    L0000: sub rsp, 0x28
    L0004: mov rax, rdx
    L0007: mov rdx, [r8]
    L000a: mov ecx, [r8+8]
    L000e: cmp r9d, ecx
    L0011: jae short L002d
    L0013: cmp r9d, ecx
    L0016: ja short L003c
    L0018: sub ecx, r9d
    L001b: mov r8d, r9d
    L001e: lea rdx, [rdx+r8*2]
    L0022: mov [rax], rdx
    L0025: mov [rax+8], ecx
    L0028: add rsp, 0x28
    L002c: ret
    L002d: xor edx, edx
    L002f: mov [rax], rdx
    L0032: xor edx, edx
    L0034: mov [rax+8], edx
    L0037: add rsp, 0x28
    L003b: ret
    L003c: call 0x00007ffb1bc85778
    L0041: int3

C.M2(System.ReadOnlySpan`1<Char>, Int32)
    L0000: mov rax, rdx
    L0003: mov rdx, [r8]
    L0006: mov ecx, [r8+8]
    L000a: cmp r9d, ecx
    L000d: ja short L0020
    L000f: sub ecx, r9d
    L0012: mov r8d, r9d
    L0015: lea rdx, [rdx+r8*2]
    L0019: mov [rax], rdx
    L001c: mov [rax+8], ecx
    L001f: ret
    L0020: xor edx, edx
    L0022: mov [rax], rdx
    L0025: xor edx, edx
    L0027: mov [rax+8], edx
    L002a: ret
```

`ReadOnlySpan<T>.Slice` contains the following guard clause:
https://github.com/dotnet/runtime/blob/e4fb7303f7be173f5434d711d42b4bba706d0ad0/src/libraries/System.Private.CoreLib/src/System/ReadOnlySpan.cs#L332-L339
In M1, we've guaranteed we're only calling `span.Slice(i)` if `i` is less than `span.Length`, but the JIT doesn't see that it can then eliminate the branch in `Slice` for when `i` is greater than `span.Length`.

In M2, the branch does get eliminated, presumably since it exactly matches the condition used by `Slice`.

cc: @AndyAyersMS

**Comment by EgorBo** [comment]
![image](https://user-images.githubusercontent.com/523221/153935296-191c7692-d2ae-46f1-ae11-64993134f4b9.png)

@AndyAyersMS from my understanding it could be handled with your prototype https://github.com/dotnet/runtime/pull/46257#issuecomment-749687576 where conditions have different VN but one implies another.

**Comment by JulieLeeMSFT** [comment]
@AndyAyersMS or @EgorBo will look into it, but we don't know if we might be able to resolve this issue.

**Comment by AndyAyersMS** [comment]
Hopefully addressed by https://github.com/dotnet/runtime/pull/72979 -- need to verify.

**Comment by AndyAyersMS** [comment]
> Hopefully addressed by #72979 -- need to verify.

Nope, not yet. These are unsigned compares and #72979 only handles signed cases.
